pinot-core/src/main/java/org/apache/pinot/core/routing/RoutingTable.java

```diff
@@ -25,25 +25,39 @@
 
 public class RoutingTable {
   private final Map<ServerInstance, List<String>> _serverInstanceToSegmentsMap;
+  private final Map<String, ServerInstance> _segmentToServerMap;
   private final List<String> _unavailableSegments;
   private final int _numPrunedSegments;
+  private Map<String, Integer> _partitionIdMap;
 
-  public RoutingTable(Map<ServerInstance, List<String>> serverInstanceToSegmentsMap, List<String> unavailableSegments,
-      int numPrunedSegments) {
+  public RoutingTable(Map<ServerInstance, List<String>> serverInstanceToSegmentsMap,
+      Map<String, ServerInstance> segmentToServerMap,
+      List<String> unavailableSegments,
+      int numPrunedSegments, Map<String, Integer> partitionIdMap) {
     _serverInstanceToSegmentsMap = serverInstanceToSegmentsMap;
+    _segmentToServerMap = segmentToServerMap;
     _unavailableSegments = unavailableSegments;
     _numPrunedSegments = numPrunedSegments;
+    _partitionIdMap = partitionIdMap;
   }
 
   public Map<ServerInstance, List<String>> getServerInstanceToSegmentsMap() {
     return _serverInstanceToSegmentsMap;
   }
 
+  public Map<String, ServerInstance> getSegmentToServerMap() {
+    return _segmentToServerMap;
+  }
+
   public List<String> getUnavailableSegments() {
     return _unavailableSegments;
   }
 
   public int getNumPrunedSegments() {
     return _numPrunedSegments;
   }
+
+  public Map<String, Integer> getSegmentToPartitionMap() {
+    return _partitionIdMap;
+  }
```

Comment:
it would be nice if we have a typical 2-stage partition architecture 

a 1-stage partition includes
1. partition key column(s)
2. partition method
3. partition count
for example:
```
class PartitionInstance implements Partitioning {
  List<ServerInstance> _serverInstances;  
}
class PartitionTable implements Partitioning {
  Map<String, List<Partitioning>> _partitionIdToInstanceListMapping;
}
```
and the RoutingTable is a special 2-stage partition 
```
class RoutingTable implements Partitioning {
  Map<String, List<Partitioning>> _partitionIdToSegmentListMapping;
}
```
for example the following list of segments
```
S1, P1, on Server1
S2, P1, on Server1
S3, P1, on Server2
S4, P2, on Server3
S5, P2, on Server4
S6, P2, on Server4
S7, P2, on Server5
```
The Routing Table will look like
```
_partitionIdToSegmentListMapping: 
{
  "P1": [ {"S1": ["Server1"], "S2": ["Server1"], "S3": ["Server2"]} ],
  "P2": [ {"S4": ["Server3"], "S5": ["Server4"], "S6": ["Server4"], "S7": ["Server5"] } ]
}
```

Reply:
rethink about this a bit. it seems to me that routingTable should be straight forward from how the routing is being produced and dispatch. thus
- the server --> partition --> segment might be better suited for this. 
- the partition --> segment --> server route is easier to construct based on the info routing manager has

Reply:
^ btw this operates under the assumption that partition-to-instance assignment for the same replica group only returns a single server (contradict to the current instance partition API which can return multiple)
it is more similar to the real-time table